agora/footfall_ai/src/main.py

```diff
@@ -0,0 +1,141 @@
+import os
+import time
+import collections
+import cv2
+import numpy as np
+import torch
+from flask import Flask, render_template, Response, request
+from ultralytics import YOLO
+from ultralytics.solutions import ObjectCounter
+from video_capture import VideoCapture
+
+# Constants
+MODEL_PATH = "C:\\Users\\fcabrera\\Downloads\\jumpstart-apps\\agora\\footfall_ai\\src\\models\\yolov8n.pt"
+RTSP_URL = os.getenv("RTSP_URL", "rtsp://localhost:554/stream")
+FRAME_RATE = 25
+CLASSES_TO_COUNT = [0]  # person is class 0 in the COCO dataset
+
+# Initialize Flask app
+app = Flask(__name__)
+
+# Global variables
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+det_model = YOLO(MODEL_PATH).to(device)
+vs = None
+fps = 0
+counter = None
+line_points = [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0)]
+
+print(f"Using device: {device}")
+print(f"Model loaded on: {next(det_model.model.parameters()).device}")
+
+def initialize_counter():
+    global counter
+    counter = ObjectCounter(
+        view_img=False,
+        reg_pts=line_points,
+        names=det_model.names,
+        draw_tracks=True,
+        line_thickness=2,
+        view_in_counts=False,
+        view_out_counts=False
+    )
+
+def gen_frames_with_source():
+    global fps, vs
+    last_time = time.time()
+    frame_count = 0
+
+    while True:
+        if(vs is None):
+            continue
+
+        frame,success = vs.read()
+        start_time = time.time()
+        if not success:
+            break
+
+        frame_count += 1
+        current_time = time.time()
+        if current_time - last_time >= 1:
+            fps = frame_count / (current_time - last_time)
+            frame_count = 0
+            last_time = current_time
+
+        cv2.putText(frame, f"FPS: {fps:.2f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (255, 255, 0), 2, cv2.LINE_AA)
+
+        if all(point == (0, 0) for point in line_points):
+            ret, frame = cv2.imencode('.jpg', frame)
+        else:
+            ret, frame = run_inference(frame)
```

Comment:
Since run_inference takes the most time out of all the preprocessing and post processing steps, we could also create a separate thread only for the inference part that would read from the captured frames queue, run inference and put in the inferenced frames queue. This is what I had tried for the old code:
https://github.com/niachary/jumpstart-agora-apps/pull/1

Reply:
@niachary  - I have another variant of this with one thread for inference and another one for capturing.. Performance wise was quite the same.   Can you please try your approach and see if its improving?

Reply:
If adding a separate thread for inference is not adding any benefit then this the approach we have is sufficient for now? Let me know what you think.

Reply:
You can try - Maybe it was my implementation, so if you know how to do it, maybe its worth to try it

Reply:
The implementation for multi threading inference thread is here if you want to take a look:
https://github.com/niachary/jumpstart-agora-apps/pull/1